sql: fix check to determine if declarative schema changer can be used

rafiss opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/141810
Release note (bug fix): Fixed a bug that would prevent CREATE TRIGGER and DROP TRIGGER statements from working if the autocommit_before_ddl setting was enabled, and if the statement was either sent as a prepared statement or as part of a batch of multiple statements.

rafiss committed:
sql: fix check to determine if declarative schema changer can be used

Release note (bug fix): Fixed a bug that would prevent CREATE TRIGGER
and DROP TRIGGER statements from working if the autocommit_before_ddl
setting was enabled, and if the statement was either sent as a prepared
statement or as part of a batch of multiple statements.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142202)
<!-- Reviewable:end -->

spilchen reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142202#-:-OKS3n60DgKfiBuGITP8:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

rafiss committed:
sql/schemachanger: change where telemetry counter is incremented

This ensures that the counter is not incremented if the statement is not
valid.

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for the fix!

> Reviewed 6 of 6 files at r3, 10 of 10 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142202#-:-OKYFvhBBYNQSO-cr8FL:b-72sm86)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

rafiss commented:
TFTR!

bors r+
